pravega/pravega
# Issue 128: Feature ReaderGroups

**Change log description**
Adds reader groups so that more than one reader can consumer events from a stream with the events being spread across them.

**Purpose of the change**
See https://github.com/emccode/pravega/issues/128
This Fixes #195, #381, #382, #383

**What the code does**
The code uses a StateSynchronizer to store the segments that are assigned to the group. This is kept up to date, by every time a reader finishes a segment, it records the following segments to the shared state. Once all the predecessors to a given segment are completed it is moved into a list of unassigned segments. Once there any reader can pick it up. Rebalancing and failover are achieved by having readers put their segments back into the unassigned list so that another reader may pick them up. The algorithm followed is described in #195  

**How to verify it**
Run the StartConsumer tests with multiple processes and write events to a stream. They should go to all of the active readers.
